components/console_cmd_mqtt/console_mqtt.c

```diff
@@ -0,0 +1,222 @@
+
+        }
+        break;
+    default:
+        ESP_LOGI(TAG, "Other event id:%d", event->event_id);
+        break;
+    }
+}
+
+static int do_mqtt_cmd(int argc, char **argv)
+{
+    int msg_id;
+    static esp_mqtt_client_handle_t client_hndl = NULL;
+    int nerrors = arg_parse(argc, argv, (void **)&mqtt_args);
+    if (nerrors != 0) {
+        arg_print_errors(stderr, mqtt_args.end, argv[0]);
+        return 1;
+    }
+
+    if (mqtt_args.connect->count > 0) {
+
+        printf("Connect command received:\n");
+        esp_mqtt_client_config_t mqtt_cfg = {
+            .broker.address.uri = CONFIG_MQTT_BROKER_URL
+        };
+
+        printf("broker: %s\n", mqtt_cfg.broker.address.uri);
+
+        client_hndl = esp_mqtt_client_init(&mqtt_cfg);
+        if (client_hndl == NULL) {
+            printf("ERROR: Client init\n");
+            return 1;
+        }
+        /* The last argument may be used to pass data to the event handler, in this example mqtt_event_handler */
+        esp_mqtt_client_register_event(client_hndl, ESP_EVENT_ANY_ID, mqtt_event_handler, NULL);
+        esp_mqtt_client_start(client_hndl);
+
+    } else if (mqtt_args.publish->count > 0) {
+
+        printf("Publish command received:\n");
+        if ((mqtt_args.topic->count <= 0) || (mqtt_args.data->count <= 0)) {
+            printf(PUBLISH_HELP_MSG);
+        }
+        printf("  Topic: %s\n", mqtt_args.topic->sval[0]);
+
+        msg_id = esp_mqtt_client_publish(client_hndl,
+                                         mqtt_args.topic->sval[0],
+                                         mqtt_args.data->sval[0],
+                                         0, 1, 0);
+        ESP_LOGI(TAG, "Publish successful, msg_id=%d, topic=%s, data=%s",
+                 msg_id, mqtt_args.topic->sval[0], mqtt_args.data->sval[0]);
+
+    } else if (mqtt_args.subscribe->count > 0) {
+        if (mqtt_args.topic->count <= 0) {
+            printf(SUBSCRIBE_HELP_MSG);
+        }
+        char *topic = (char *)mqtt_args.topic->sval[0];
+        printf("subscribe command received:\n");
+        printf("  Topic: %s\n", topic);
+        msg_id = esp_mqtt_client_subscribe(client_hndl, topic, 0);
+        ESP_LOGI(TAG, "sent subscribe successful, msg_id=%d", msg_id);
+
+    } else if (mqtt_args.unsubscribe->count > 0) {
+        if (mqtt_args.topic->count <= 0) {
+            printf(UNSUBSCRIBE_HELP_MSG);
+        }
+        printf("unsubscribe command received:\n");
+        printf("  Topic: %s\n", mqtt_args.topic->sval[0]);
+        msg_id = esp_mqtt_client_unsubscribe(client_hndl, mqtt_args.topic->sval[0]);
+        ESP_LOGI(TAG, "sent unsubscribe successful, msg_id=%d", msg_id);
+
+    } else if (mqtt_args.list_subscriptions->count > 0) {
+        printf("TBD: list_subscriptions command\n");
+
+    } else if (mqtt_args.disconnect->count > 0) {
+        printf("Disconnect command received:\n");
+        esp_mqtt_client_disconnect(client_hndl);
```

Comment:
should we also disable the auto-reconnection feature? Because the client will probably connect again shortly

Reply:
Stopping the client instead of just disconnect.